File: android/src/main/java/com/monri/flutter/FlutterConfirmPaymentParams.java
package com.monri.flutter;

import com.monri.android.model.Card;
import com.monri.android.model.ConfirmPaymentParams;
import com.monri.android.model.CustomerParams;
import com.monri.android.model.GooglePayPayment;
import com.monri.android.model.MonriApiOptions;
import com.monri.android.model.PaymentMethodParams;
import com.monri.android.model.SavedCard;
import com.monri.android.model.TransactionParams;

import java.util.Map;

public class FlutterConfirmPaymentParams {

    private boolean developmentMode;
    private String authenticityToken;
    private String clientSecret;

    private FlutterCard card;
    private FlutterSavedCard savedCard;

    private FlutterTransactionParams transactionParams;

    private FlutterGooglePay googlePayData;

    private FlutterConfirmPaymentParams(boolean developmentMode, String authenticityToken, String clientSecret,
                                        FlutterCard card, FlutterSavedCard savedCard, FlutterTransactionParams transactionParams, FlutterGooglePay googlePayData) {
        this.developmentMode = developmentMode;
        this.authenticityToken = authenticityToken;
        this.clientSecret = clientSecret;
        this.card = card;
        this.savedCard = savedCard;
        this.transactionParams = transactionParams;
        this.googlePayData = googlePayData;
    }

    private PaymentMethodParams paymentMethodParams() {
        if (card != null) {
            return new Card(card.pan, card.month, card.year, card.cvv).setTokenizePan(card.tokenizePan)
                                                                      .toPaymentMethodParams();
        } else if (savedCard != null) {
            return new SavedCard(savedCard.panToken, savedCard.cvv).toPaymentMethodParams();
        } else if (googlePayData != null) {
            return new GooglePayPayment(GooglePayPayment.Provider.GOOGLE_PAY).toPaymentMethodParams();
        } else {
            return new SavedCard(savedCard.panToken, savedCard.cvv).toPaymentMethodParams();
        }
    }

    private TransactionParams transactionParams() {
        final CustomerParams customerParams = new CustomerParams()
                .setAddress(transactionParams.address)
                .setFullName(transactionParams.fullName)
                .setCity(transactionParams.city)
                .setZip(transactionParams.zip)
                .setPhone(transactionParams.phone)
                .setCountry(transactionParams.country)
                .setEmail(transactionParams.email);

        final TransactionParams rv = TransactionParams.create()
                                                      .set("custom_params", transactionParams.customParams)
                                                      .set(customerParams);

        if (Boolean.TRUE.equals(transactionParams.moto)) {
            rv.set("moto", "true");
        }

        return rv;
    }

    ConfirmPaymentParams confirmPaymentParams() {
        return ConfirmPaymentParams.create(clientSecret, paymentMethodParams(), transactionParams());
    }

    MonriApiOptions monriApiOptions() {
        return MonriApiOptions.create(authenticityToken, developmentMode);
    }

    MonriApiOptions monriApiOptions(String authenticityToken, boolean developmentMode) {
        return MonriApiOptions.create(authenticityToken, developmentMode);
    }

    public FlutterGooglePay getGooglePayData() {
        return googlePayData;
    }

    @SuppressWarnings("unchecked")
    private static FlutterCard card(Map<String, Object> cardMap) {

        String number = (String) cardMap.get("pan");
        String cvv = (String) cardMap.get("cvv");
        boolean tokenizePan = (Boolean) cardMap.get("tokenize_pan");
        int expYear = (Integer) cardMap.get("expiry_year");
        int expMonth = (Integer) cardMap.get("expiry_month");
        return new FlutterCard(number, cvv, expYear, expMonth, tokenizePan);
    }

    @SuppressWarnings("unchecked")
    private static FlutterSavedCard savedCard(Map<String, Object> cardMap) {
        return new FlutterSavedCard(
                (String) cardMap.get("pan_token"),
                (String) cardMap.get("cvv")
        );
    }

    @SuppressWarnings("unchecked")
    public static FlutterConfirmPaymentParams forCard(Map<String, Object> map) {
        return create(map, card((Map<String, Object>) map.get("card")), null, null);
    }

    @SuppressWarnings("unchecked")
    public static FlutterConfirmPaymentParams forSavedCard(Map<String, Object> map) {
        return create(map, null, savedCard((Map<String, Object>) map.get("saved_card")), null);
    }

    public static FlutterConfirmPaymentParams forGooglePay(Map<String, Object> map) {
        final int gPayType;
        final int gPayTheme;
        final int gPayCornerRadius;

        Object typeObj = map.get("gPayButtonType");
        if (typeObj != null) {
            gPayType = (int) typeObj;
        } else {
            gPayType = 1; // or some default value
        }

        Object themeObj = map.get("gPayButtonTheme");
        if (typeObj != null) {
            gPayTheme = (int) themeObj;
        } else {
            gPayTheme = 1; // or some default value
        }

        Object cornerRadiusObj = map.get("gPayCornerRadius");
        if (typeObj != null) {
            gPayCornerRadius = (int) cornerRadiusObj;
        } else {
            gPayCornerRadius = 100; // or some default value
        }

        return create(map, null, null, new FlutterGooglePay(gPayTheme, gPayType, gPayCornerRadius));
    }

    @SuppressWarnings("unchecked")
    private static FlutterConfirmPaymentParams create(Map<String, Object> request, FlutterCard card,
                                                      FlutterSavedCard savedCard, FlutterGooglePay googlePayData) {

        String authenticityToken = (String) request.get("authenticity_token");
        String clientSecret = (String) request.get("client_secret");
        boolean developmentMode = (boolean) request.get("is_development_mode");
        Map<String, Object> transactionParamsJSON = (Map<String, Object>) request.get("transaction_params");

        FlutterTransactionParams transactionParams = new FlutterTransactionParams(
                (String) transactionParamsJSON.get("order_info"),
                (String) transactionParamsJSON.get("email"),
                (String) transactionParamsJSON.get("full_name"),
                (String) transactionParamsJSON.get("address"),
                (String) transactionParamsJSON.get("city"),
                (String) transactionParamsJSON.get("zip"),
                (String) transactionParamsJSON.get("country"),
                (String) transactionParamsJSON.get("custom_params"),
                (String) transactionParamsJSON.get("phone"),
                (Boolean) transactionParamsJSON.get("moto"));

        return new FlutterConfirmPaymentParams(developmentMode, authenticityToken, clientSecret, card, savedCard,
                                               transactionParams, googlePayData);
    }

    static class FlutterTransactionParams {

        String orderInfo;
        String email;
        String fullName;
        String address;
        String city;
        String zip;
        String phone;
        String country;
        String customParams;
        Boolean moto;

        FlutterTransactionParams(String orderInfo,
                                 String email,
                                 String fullName,
                                 String address,
                                 String city,
                                 String zip,
                                 String country,
                                 String customParams,
                                 String phone,
                                 Boolean moto) {
            this.orderInfo = orderInfo;
            this.email = email;
            this.fullName = fullName;
            this.address = address;
            this.city = city;
            this.zip = zip;
            this.country = country;
            this.customParams = customParams;
            this.phone = phone;
            this.moto = moto;
        }
    }

    static class FlutterCard {

        String pan;
        String cvv;
        int year;
        int month;
        boolean tokenizePan;

        FlutterCard(String pan, String cvv, int year,
                    int month, boolean tokenizePan) {
            this.pan = pan;
            this.cvv = cvv;
            this.tokenizePan = tokenizePan;
            this.year = year;
            this.month = month;
        }
    }

    static class FlutterSavedCard {

        String panToken;
        String cvv;

        FlutterSavedCard(String panToken, String cvv) {
            this.panToken = panToken;
            this.cvv = cvv;
        }
    }

    static class FlutterGooglePay {

        final int gPayTheme;
        final int gPayButtonType;
        final int gPayCornerRadius;

        FlutterGooglePay(final int gPayTheme, final int gPayButtonType, final int gPayCornerRadius) {
            this.gPayTheme = gPayTheme;
            this.gPayButtonType = gPayButtonType;
            this.gPayCornerRadius = gPayCornerRadius;
        }
    }
}

File: ios/Classes/FlutterConfirmPaymentParams.swift
//
//  FlutterConfirmPaymentParams.swift
//  MonriPayments
//
//  Created by Jasmin Suljic on 18/11/2020.
//

import Foundation
import Monri
//
//  FlutterConfirmPaymentParams.swift
//  MonriPayments
//
//  Created by Jasmin Suljic on 18/11/2020.
//

import Foundation

class FlutterConfirmPaymentParams {
    let developmentMode: Bool
    let authenticityToken: String
    let clientSecret: String
    let card: FlutterCard?
    let savedCard: FlutterSavedCard?
    let transactionParams: FlutterTransactionParams
    
    init(_ developmentMode: Bool,_ authenticityToken: String,_ clientSecret: String,_ card: FlutterCard?,_ savedCard: FlutterSavedCard?,_ transactionParams: FlutterTransactionParams) {
        self.developmentMode = developmentMode
        self.authenticityToken = authenticityToken
        self.clientSecret = clientSecret
        self.card = card
        self.savedCard = savedCard
        self.transactionParams = transactionParams
    }
    
    func monriApiOptions() ->MonriApiOptions {
        return MonriApiOptions(authenticityToken: authenticityToken, developmentMode: developmentMode)
    }
    
    private static func card(cardMap: Dictionary<String, AnyObject?>) -> FlutterCard {
        let number: String = cardMap["pan"] as! String
        let cvv: String = cardMap["cvv"] as! String
        let tokenizePan: Bool = cardMap["tokenize_pan"] as! Bool
        let expYear: Int = cardMap["expiry_year"] as! Int
        let expMonth: Int = cardMap["expiry_month"] as! Int
        return FlutterCard(number, cvv, expYear, expMonth, tokenizePan)
    }
    
    private func paymentMethodParams() -> PaymentMethodParams {
        if let card = card {
            return Card(number: card.pan, cvc: card.cvv, expMonth: card.month, expYear: card.year, tokenizePan: card.tokenizePan).toPaymentMethodParams()
        }
        
        return SavedCard(panToken: savedCard!.panToken, cvc: savedCard!.cvv).toPaymentMethodParams()
    }
    
    private func buildMonriTransactionParams() -> TransactionParams {
        let customerParams: CustomerParams = CustomerParams(email: transactionParams.email,
                                                            fullName: transactionParams.fullName,
                                                            address: transactionParams.address,
                                                            city: transactionParams.city,
                                                            zip: transactionParams.zip,
                                                            phone: transactionParams.phone,
                                                            country: transactionParams.country
        )
        
        return TransactionParams.create()
            .set(customerParams:customerParams)
            .set("custom_params", transactionParams.customParams)
    }
    
    
    private static  func savedCard(cardMap: Dictionary<String, AnyObject?>) -> FlutterSavedCard {
        return FlutterSavedCard(
            cardMap["pan_token"] as! String,
            cardMap["cvv"] as! String
        )
    }
    
    func confirmPaymentParams() -> ConfirmPaymentParams {
        return ConfirmPaymentParams(paymentId: clientSecret, paymentMethod: paymentMethodParams(), transaction: buildMonriTransactionParams())
    }
    
    static func create(request: Dictionary<String, AnyObject?>, card: FlutterCard?, savedCard: FlutterSavedCard?) -> FlutterConfirmPaymentParams {
        let authenticityToken: String = request["authenticity_token"] as! String
        let clientSecret: String = request["client_secret"] as! String
        let developmentMode: Bool = request["is_development_mode"] as! Bool
        let transactionParamsJSON: Dictionary<String, AnyObject?> = request["transaction_params"] as! Dictionary<String, AnyObject?>
        
        let transactionParams: FlutterTransactionParams  = FlutterTransactionParams(
            orderInfo: transactionParamsJSON["order_info"] as? String,
            email: transactionParamsJSON["email"] as? String,
            fullName: transactionParamsJSON["full_name"] as? String,
            address: transactionParamsJSON["address"] as? String,
            city: transactionParamsJSON["city"] as? String,
            zip: transactionParamsJSON["zip"] as? String,
            phone: transactionParamsJSON["phone"] as? String,
            country: transactionParamsJSON["country"] as? String,
            customParams:transactionParamsJSON["custom_params"] as? String
        )
        
        return FlutterConfirmPaymentParams(developmentMode, authenticityToken, clientSecret, card, savedCard, transactionParams)
    }
    
    static func forCard(request: Dictionary<String, AnyObject?>) -> FlutterConfirmPaymentParams {
        return Self.create(request: request,
                           card: Self.card(cardMap: request["card"] as! Dictionary<String, AnyObject?>),
                           savedCard: nil
        )
    }
    
    static func forSavedCard(request: Dictionary<String, AnyObject?>) -> FlutterConfirmPaymentParams {
        return Self.create(request: request,
                           card: nil,
                           savedCard: Self.savedCard(cardMap: request["saved_card"] as! Dictionary<String, AnyObject?>)
        )
    }
}

class FlutterCard {
    let pan: String
    let cvv: String
    let year: Int
    let month: Int
    let tokenizePan: Bool
    
    init(_ pan: String,_ cvv: String,_ year: Int,_ month: Int,_ tokenizePan: Bool) {
        self.pan = pan
        self.cvv = cvv
        self.year = year
        self.month = month
        self.tokenizePan = tokenizePan
    }
}

class FlutterSavedCard {
    let panToken: String
    let cvv: String
    init(_ panToken: String, _ cvv:String) {
        self.panToken = panToken
        self.cvv = cvv
    }
}

class FlutterTransactionParams {
    let orderInfo: String?
    let email: String?
    let fullName: String?
    let address: String?
    let city: String?
    let zip: String?
    let phone: String?
    let country: String?
    let customParams: String?
    
    init(orderInfo: String?,
         email: String?,
         fullName: String?,
         address: String?,
         city: String?,
         zip: String?,
         phone: String?,
         country: String?,
         customParams: String?
    ) {
        self.orderInfo = orderInfo
        self.email = email
        self.fullName = fullName
        self.address = address
        self.city = city
        self.zip = zip
        self.phone = phone
        self.country = country
        self.customParams = customParams
    }
}

